add forking functionality

hamilton opened the pull request:
![screen shot 2018-10-02 at 10 42 44 pm](https://user-images.githubusercontent.com/95735/46392017-82664500-c694-11e8-893a-5575996107c7.png)


I think this could probably be refined, but figured I'd put this PR together to get some feedback.

Closes #836 and #915. Specifically, it:

- [x] adds a `forked_from` field in the `Notebook` model, which is a `ForeignKey` on the `NotebookRevision`. This allows the history traceability back to the notebook revision that spawned the new notebook. This does _not_ address adding a `parent_id` to each revision at the moment. I'm not sure what having a `parent_id` actually buys us in terms of workflow.
- [x] in the Notebook API, if you add `forked_from: <id>` to the post request, it will add this field, otherwise defaults to `null`. Adding a `forked_from` id that does not exist should throw an error.
- [x] now you can fork from the notebook interstitial directly, and this will be the only supported way to fork for this PR. Future PRs will add further forking support in menus etc.
- [x] on the revisions page you see if / who / which notebook the forked notebook came from.
- [x] write a few server tests to ensure forking is handled correctly.

hamilton commented on server/notebooks/migrations/0002_notebook_forked_from.py:
```diff
@@ -0,0 +1,19 @@
+# Generated by Django 2.1 on 2018-10-03 01:54
```
@wlach I should probably remove this, yeah? There is another migration added to the repository so I added.

wlach replied:
Fine to leave the comment in.

hamilton commented on server/tests/test_notebook_api.py:
```diff
@@ -105,3 +108,47 @@ def test_delete_notebook_owner(fake_user, test_notebook, client):
+    assert resp.status_code == 403
+
+
+def test_incorrect_fork_notebook(client, fake_user, incorrect_notebook_fork_post_blob):
```
This set of assertions feels incomplete.

wlach commented on server/tests/test_notebook_api.py:
```diff
@@ -105,3 +108,47 @@ def test_delete_notebook_owner(fake_user, test_notebook, client):
+    client.force_authenticate(user=fake_user)
+    resp = client.post(reverse('notebooks-list'), notebook_fork_post_blob)
+    assert resp.status_code == 201
+    assert resp.json()['forked_from'] == test_notebook.revisions.get().id
```
I would like to see another test that deletes the notebook that another was forked from, verifying that the forked notebook still exists.

hamilton replied:
added this test in a commit below.

wlach commented on src/actions/actions.js:
```diff
@@ -269,6 +269,13 @@ function getNotebookSaveRequestOptions(state) {
   }
+  const revisionID = state.notebookInfo.revision_id
+  const userOwns = state.notebookInfo.user_can_save
+  const forkedFrom = revisionID || undefined
```
the `|| undefined` part seems redundant?

wlach commented on src/actions/actions.js:
```diff
@@ -269,6 +269,13 @@ function getNotebookSaveRequestOptions(state) {
+  const revisionID = state.notebookInfo.revision_id
+  const userOwns = state.notebookInfo.user_can_save
+  const forkedFrom = revisionID || undefined
+  // only set forked_from if user doesn't own this notebook.
```
Hmm, I feel like we might want to keep track of the history even if the user is forking their own notebook?

hamilton replied:
good point

hamilton replied:
Going to generalize this a little bit so we can make it easy for a user to fork their notebook from a menu item, etc.

hamilton commented:
@wlach I believe I addressed the your comments w/ this most recent set of commits - feel free to take a look at your convenience and let me know if this looks good.

wlach commented:
Looks good now! I'll merge.
